src/Features/LanguageServer/Microsoft.CommonLanguageServerProtocol.Framework/AbstractLanguageServer.cs

```diff
@@ -0,0 +1,327 @@
+        var lspServices = GetLspServices();
+
+        var handlerProvider = GetHandlerProvider();
+        var queue = new RequestExecutionQueue<RequestContextType>(_logger, handlerProvider);
+        queue.RequestServerShutdown += RequestExecutionQueue_Errored;
+
+        queue.Start(lspServices);
+
+        return queue;
+    }
+
+    protected IRequestExecutionQueue<RequestContextType> GetRequestExecutionQueue()
+    {
+        if (_queue is null)
+            _queue = ConstructRequestExecutionQueue();
+
+        return _queue;
+    }
+
+    /// <summary>
+    /// Wrapper class to hold the method and properties from the <see cref="AbstractLanguageServer{RequestContextType}"/>
+    /// that the method info passed to streamjsonrpc is created from.
+    /// </summary>
+    private class DelegatingEntryPoint
+    {
+        private readonly string _method;
+        private readonly AbstractLanguageServer<RequestContextType> _target;
+
+        public DelegatingEntryPoint(string method, AbstractLanguageServer<RequestContextType> target)
+        {
+            _method = method;
+            _target = target;
+        }
+
+        public async Task NotificationEntryPointAsync<TRequestType>(TRequestType request, CancellationToken cancellationToken) where TRequestType : class
+        {
+            CheckServerState();
+            var queue = _target.GetRequestExecutionQueue();
+            var lspServices = _target.GetLspServices();
+
+            _ = await queue.ExecuteAsync<TRequestType, VoidReturn>(request, _method, lspServices, cancellationToken).ConfigureAwait(false);
+        }
+
+        public async Task ParameterlessNotificationEntryPointAsync(CancellationToken cancellationToken)
+        {
+            CheckServerState();
+            var queue = _target.GetRequestExecutionQueue();
+            var lspServices = _target.GetLspServices();
+
+            _ = await queue.ExecuteAsync<VoidReturn, VoidReturn>(VoidReturn.Instance, _method, lspServices, cancellationToken).ConfigureAwait(false);
+        }
+
+        public async Task<TResponseType?> EntryPointAsync<TRequestType, TResponseType>(TRequestType request, CancellationToken cancellationToken) where TRequestType : class
+        {
+            CheckServerState();
+            var queue = _target.GetRequestExecutionQueue();
+            var lspServices = _target.GetLspServices();
+
+            var result = await queue.ExecuteAsync<TRequestType, TResponseType>(request, _method, lspServices, cancellationToken).ConfigureAwait(false);
+
+            return result;
+        }
+
+        private void CheckServerState()
+        {
+            if (_target.IsInitialized)
+                throw new InvalidOperationException($"'initialize' has not been called.");
+        }
+    }
+
+    public virtual async Task ShutdownAsync()
+    {
+        _shuttingDown = true;
+
+        await ShutdownRequestExecutionQueueAsync().ConfigureAwait(false);
+    }
+
+    public virtual async Task ExitAsync()
```

Comment:
shutdown / exit logic is a quite confusing as to who calls who/ who owns what - it might make sense if these are solely called by the lifecycle manager to make these an explicit implementation of a lifecycle interface.

Also - do people who want custom logic implement these virtual methods or the lifecycle manager?  We should pick one.